### PR TITLE
FEATURE: allow multiple secrets for Discourse SSO provider

### PR DESCRIPTION
**Pull request by majakomel**
https://meta.discourse.org/t/supporting-multiple-sso-secrets-with-discourse-as-provider/96693

### CODE REVIEW

**Comment by SamSaffron** [comment]
Awesome feature @majakomel  can you do a quick screenshot of how this looks with your change?

**Comment by majakomel** [comment]
@SamSaffron I'd like to improve the flow from the user perspective, will consult @jjaffeux tomorrow and and then post screenshots

**Comment by gschlager** [comment]
I think it would be better to use only `sso_provider_secrets` and not fallback to `sso_secret`. Maybe migrate the value of `sso_secret` as `*|old_sso_secret`?

**Comment by jjaffeux** [comment]
Just a note here, I had a call with @majakomel and we will modify the UI here, the basic of the feature are here but we will try to achieve a better experience.

**Comment by majakomel** [comment]
This is how sso provider secrets site setting looks like:

screen shot 2018-10-05 at 14 11 00

File: lib/single_sign_on.rb [hunk 1]
```diff
@@ -83,6 +88,18 @@ def self.parse(payload, sso_secret = nil)
     sso
   end
 
+  def self.provider_secret(return_sso_url)
+    provider_secrets = SiteSetting.sso_provider_secrets.split(/[\|,\n]/)
+    provider_secrets_hash = Hash[*provider_secrets]
+    return_url = URI.parse(return_sso_url).host
+
+    secret = provider_secrets_hash.select do |domain, _|
+      domain_regex = Regexp.escape(domain).gsub("\\*", '\S*')
```
Comment (SamSaffron) [inline lib/single_sign_on.rb]: Is there a bug here ... Say we allow. `*.discourse.org` will `bad-domain.discourse.org.evil.com` slip through? maybe extract some of this logic into a service object of sorts and add some tests for the security aspects?

**Comment by SamSaffron** [comment]
Thanks heaps @majakomel  👍  merging it in!

**Pull request merged**